### ISSUE TITLE
Create a feedback module

### ISSUE THREAD

**Issue by SimonBiggs** [pymedphys/pymedphys#1721]
We could create a feedback module with the purpose of gathering feedback for improving the library, information for supporting maintenance decisions, as well as building an evidence base of the impact being made by PyMedPhys to support potential grant applications in the future.

There are many ways we could design this. I have presented a few ideas below:

Create a `pymedphys-user-community` GitHub account which has push access (but not force push access) to a purpose build repo `pymedphys/feedback`. We provide the password to this account within the PyMedPhys library, we comment around it mentioning that this password is being shared in good faith, detailing why we provide it, detailing the benefits of producing public metrics, and request that the account not be misused.

Create a `pymedphys.feedback.collect` decorator which we use to decorate our front facing API. This metrics API will only collect if the package has been installed from PyPI, the test suite is not running, and it is not detectibly running on CI infrastructure.

All data collected will be saved locally, and then it will be intermittently uploaded to git repo via a git push. The username and email address for the commit will be that of the `pymedphys-user-community` account.

Data collected will include a random UUID created during PyMedPhys first use on any given machine. As well as a call count to each of the front facing API. If an error message is raised, the decorator will read the traceback and if a PyMedPhys function was at the core of the issue then the error type and internal PyMedPhys function names and line numbers will be recorded. The PyMedPhys version number, as well as dependency version numbers will also be recorded.

We can also provide a full stack trace and feedback submission function, `pymedphys.feedback.error_report`. This will create an error report with the most recent stack trace as well as an opportunity for the user to provide a description of what they were trying to do. We can have the a description detailing how to use of this function be appended onto any error messages that occur when utilising a function that has been decorated by `pymedphys.feedback.collect`. When the user calls the function, a full print out of the report will be provided to the user prior to submission. We would need to make sure that only PyMedPhys code is ever shown in this report and no input or output variable values.

On first use of PyMedPhys a warning message is presented mentioning that metrics will be collected, as well as briefly detailing what they will be used for and where to get more detailed information. We will have an opt-out function that when called, permanently turns off all feedback metric collection and submission functionality on that machine.

Thoughts?

It also might be helpful to provide a means to sparingly and intermittently request users to fill out a survey. In this survey they have the opportunity to provide positive and negative feedback. They also have the opportunity to choose to provide their contact details, and be asked questions along the lines of, "Might you or your organisation be interested in partnering with PyMedPhys for PyMedPhys grant submissions in the future?" "Might you or your organisation be willing to provide any co-contributions, in-kind or otherwise as a part of these grant application?"

**Comment by sjswerdloff** [comment]
I worry about the abuse by non users (of the submission mechanism and the access it provides, even if it is just a denial of service attack).
But the rest looks good to me.

I would like the stats gathered by the site to be readily observable to the site, e.g. by including basic review tools in PyMedPhys (even if its just a text dump or bar chart).

I'm not a fan of telemetry when I don't know exactly what is contained.  I am much happier when I see the content...

**Comment by SimonBiggs** [comment]
> I would like the stats gathered by the site to be readily observable to the site, e.g. by including basic review tools in PyMedPhys (even if its just a text dump or bar chart).

Yup, definitely. It needs to be above and beyond above board. Potentially we can build in the stats analysis functions for both your locally stored records but also for the global cohort. We could potentially even have a "how-to" example notebook that shows the various stats collected and provides a way to investigate it and extract helpful insights.

> I worry about the abuse by non users (of the submission mechanism and the access it provides, even if it is just a denial of service attack).

It would be open to DOS attack. In the event that happens we could revoke the repo access to that user and investigate an alternative. GitHub itself does have some DOS defence measures. And by not allowing force pushes, we can always revert the history ourselves to prior to the attack. 

Worst case scenario, as I see it, is that we'd have to undo the upload approach and re-solve the data submission problem.

**Comment by SimonBiggs** [comment]
Here is a good article in nature about setting up projects for the long haul and it provides some grant funding options:

https://www.nature.com/articles/d41586-019-02046-0

I know if we could get grant funding that is enough to cover a reasonable software developer salary I would be quite happy commiting to work on PyMedPhys full time and help make it be able to become one of the best versions of what it possibly might be able to become. I imagine others might also be in the same boat.
